File: internal/uuidx/uuidx_test.go
package uuidx

import (
	"strings"
	"testing"

	"github.com/google/uuid"
)

func TestEncodeDecodeRoundtripFixed(t *testing.T) {
	cases := []struct {
		name string
		in   [16]byte
		want string
	}{
		{"all-zero", [16]byte{}, strings.Repeat("0", 26)},
		{"all-FF", fillByte(0xFF), "7ZZZZZZZZZZZZZZZZZZZZZZZZZ"},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			got := Encode(c.in)
			if got != c.want {
				t.Fatalf("Encode(%x) = %q, want %q", c.in, got, c.want)
			}
			back, err := Decode(got)
			if err != nil {
				t.Fatalf("Decode(%q) error: %v", got, err)
			}
			if back != c.in {
				t.Fatalf("Decode(Encode(x)) = %x, want %x", back, c.in)
			}
		})
	}
}

// TestEncodeDecodeRoundtripRandom exercises the round-trip property over
// 10000 random byte arrays. Required by §7 of the portable-builds plan.
func TestEncodeDecodeRoundtripRandom(t *testing.T) {
	const n = 10000
	for range n {
		raw, err := randomBytes(16)
		if err != nil {
			t.Fatalf("rand: %v", err)
		}
		var in [16]byte
		copy(in[:], raw)
		s := Encode(in)
		if len(s) != 26 {
			t.Fatalf("encoded length = %d, want 26 (input %x)", len(s), in)
		}
		out, err := Decode(s)
		if err != nil {
			t.Fatalf("Decode(%q) error: %v", s, err)
		}
		if out != in {
			t.Fatalf("roundtrip mismatch: in=%x out=%x via %q", in, out, s)
		}
	}
}

func TestEncodeAlways26Chars(t *testing.T) {
	for i := range 256 {
		var in [16]byte
		in[0] = byte(i) // sweep top byte
		s := Encode(in)
		if len(s) != 26 {
			t.Fatalf("len=%d for top byte %02x", len(s), i)
		}
	}
}

func TestDecodeRejectsBadLength(t *testing.T) {
	for _, s := range []string{"", "0", strings.Repeat("0", 25), strings.Repeat("0", 27)} {
		if _, err := Decode(s); err == nil {
			t.Fatalf("Decode(%q) expected error, got nil", s)
		}
	}
}

func TestDecodeRejectsInvalidChar(t *testing.T) {
	// 'U' is excluded from Crockford base32.
	bad := strings.Repeat("0", 25) + "U"
	if _, err := Decode(bad); err == nil {
		t.Fatalf("Decode(%q) expected error, got nil", bad)
	}
}

func TestDecodeAcceptsCrockfordAliases(t *testing.T) {
	// I/L are aliases for 1, O is alias for 0.
	canonical := "10000000000000000000000010"
	for _, alias := range []string{
		"I0000000000000000000000010", // I → 1
		"L0000000000000000000000010", // L → 1
		"i0000000000000000000000010", // lowercase
		"10000O00000000000000000010", // O → 0 (no-op since already 0, but exercises map)
	} {
		want, _ := Decode(canonical)
		got, err := Decode(alias)
		if err != nil {
			t.Fatalf("Decode(%q) error: %v", alias, err)
		}
		if got != want {
			t.Fatalf("alias %q decoded to %x, want %x", alias, got, want)
		}
	}
}

func TestDecodeRejectsOversizedFirstChar(t *testing.T) {
	// First char must be 0-7; any value >7 implies more than 128 bits.
	for _, c := range []byte{'8', '9', 'A', 'Z'} {
		s := string(c) + strings.Repeat("0", 25)
		if _, err := Decode(s); err == nil {
			t.Fatalf("Decode(%q) expected error, got nil", s)
		}
	}
}

func TestNewProducesUUIDv7(t *testing.T) {
	b, err := New()
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	u, err := uuid.FromBytes(b[:])
	if err != nil {
		t.Fatalf("FromBytes: %v", err)
	}
	if u.Version() != 7 {
		t.Fatalf("version = %d, want 7", u.Version())
	}
}

// TestEncodePreservesLexOrder confirms that two byte arrays compared as
// big-endian 128-bit integers retain their order after encoding. UUIDv7's
// time-prefix relies on this for natural chronological listing in URLs.
func TestEncodePreservesLexOrder(t *testing.T) {
	pairs := []struct {
		lo, hi [16]byte
	}{
		{[16]byte{}, fillByte(0xFF)},
		{[16]byte{0x00, 0x00, 0x00, 0x01}, [16]byte{0x00, 0x00, 0x00, 0x02}},
		{[16]byte{0x7F, 0xFF, 0xFF, 0xFF}, [16]byte{0x80, 0x00, 0x00, 0x00}},
	}
	for _, p := range pairs {
		l, h := Encode(p.lo), Encode(p.hi)
		if !(l < h) {
			t.Fatalf("lex order violated: %q !< %q", l, h)
		}
	}
}

func fillByte(v byte) [16]byte {
	var b [16]byte
	for i := range b {
		b[i] = v
	}
	return b
}

File: web/src/uuidx.test.ts
import { describe, expect, test } from "bun:test";
import { bytesEqual, decode, encode, newUuidV7 } from "./uuidx";

function fillByte(v: number): Uint8Array {
  const b = new Uint8Array(16);
  b.fill(v);
  return b;
}

describe("encode/decode fixed cases", () => {
  test("all-zero round-trips through 26 zero chars", () => {
    const z = new Uint8Array(16);
    expect(encode(z)).toBe("0".repeat(26));
    expect(bytesEqual(decode("0".repeat(26)), z)).toBe(true);
  });

  test("all-FF round-trips through 7ZZ...ZZ", () => {
    const f = fillByte(0xff);
    expect(encode(f)).toBe("7ZZZZZZZZZZZZZZZZZZZZZZZZZ");
    expect(bytesEqual(decode("7ZZZZZZZZZZZZZZZZZZZZZZZZZ"), f)).toBe(true);
  });
});

describe("encode/decode round-trip property", () => {
  // Mirrors Go's TestEncodeDecodeRoundtripRandom — 10000 random byte
  // arrays must survive encode→decode with byte-for-byte identity.
  // Required by §7 of docs/plans/portable-builds.md.
  test("10000 random round-trips", () => {
    const buf = new Uint8Array(16);
    for (let i = 0; i < 10000; i++) {
      crypto.getRandomValues(buf);
      const s = encode(buf);
      expect(s.length).toBe(26);
      const back = decode(s);
      if (!bytesEqual(back, buf)) {
        throw new Error(
          `roundtrip mismatch via ${s}: in=${[...buf]} out=${[...back]}`
        );
      }
    }
  });
});

describe("encode invariants", () => {
  test("always produces 26 chars across the top byte sweep", () => {
    const b = new Uint8Array(16);
    for (let i = 0; i < 256; i++) {
      b[0] = i;
      expect(encode(b).length).toBe(26);
    }
  });

  test("preserves lex order", () => {
    const pairs: [Uint8Array, Uint8Array][] = [
      [new Uint8Array(16), fillByte(0xff)],
      [
        Uint8Array.from([0, 0, 0, 1, ...new Array(12).fill(0)]),
        Uint8Array.from([0, 0, 0, 2, ...new Array(12).fill(0)]),
      ],
      [
        Uint8Array.from([0x7f, 0xff, 0xff, 0xff, ...new Array(12).fill(0)]),
        Uint8Array.from([0x80, 0x00, 0x00, 0x00, ...new Array(12).fill(0)]),
      ],
    ];
    for (const [lo, hi] of pairs) {
      const l = encode(lo);
      const h = encode(hi);
      expect(l < h).toBe(true);
    }
  });
});

describe("decode error handling", () => {
  test("rejects bad length", () => {
    for (const s of ["", "0", "0".repeat(25), "0".repeat(27)]) {
      expect(() => decode(s)).toThrow();
    }
  });

  test("rejects invalid characters", () => {
    // U is excluded from Crockford base32.
    expect(() => decode("0".repeat(25) + "U")).toThrow();
  });

  test("rejects an oversized first character", () => {
    // First char > 7 implies more than 128 bits.
    for (const c of ["8", "9", "A", "Z"]) {
      expect(() => decode(c + "0".repeat(25))).toThrow();
    }
  });

  test("accepts Crockford aliases I/L → 1 and O → 0", () => {
    const canonical = "10000000000000000000000010";
    const want = decode(canonical);
    for (const alias of [
      "I0000000000000000000000010",
      "L0000000000000000000000010",
      "i0000000000000000000000010",
      "10000O00000000000000000010",
    ]) {
      const got = decode(alias);
      expect(bytesEqual(got, want)).toBe(true);
    }
  });
});

describe("newUuidV7", () => {
  test("returns 16 bytes with version=7 and variant=10", () => {
    const b = newUuidV7();
    expect(b.length).toBe(16);
    expect((b[6] & 0xf0) >> 4).toBe(7);
    expect((b[8] & 0xc0) >> 6).toBe(0b10);
  });

  test("two consecutive ids differ", () => {
    const a = newUuidV7();
    const b = newUuidV7();
    expect(bytesEqual(a, b)).toBe(false);
  });

  test("encoded form survives a round-trip", () => {
    for (let i = 0; i < 100; i++) {
      const b = newUuidV7();
      const s = encode(b);
      expect(s.length).toBe(26);
      expect(bytesEqual(decode(s), b)).toBe(true);
    }
  });

  test("time prefix tracks Date.now() within 1 second", () => {
    const before = Date.now();
    const b = newUuidV7();
    const after = Date.now();
    // First 48 bits = epoch ms, big-endian.
    const ms =
      b[0] * 0x10000000000 +
      b[1] * 0x100000000 +
      b[2] * 0x1000000 +
      b[3] * 0x10000 +
      b[4] * 0x100 +
      b[5];
    expect(ms).toBeGreaterThanOrEqual(before);
    expect(ms).toBeLessThanOrEqual(after);
  });
});
